### ISSUE TITLE
`llpcElfWriter.cpp` does not check returned `Result`s

### ISSUE THREAD

**Issue by kuhar** [GPUOpen-Drivers/llpc#1468]
I annotated functions returning `Result` to illustrate the issue:

```
[5/23] Building CXX object compiler/llpc/CMakeFiles/llpc.dir/util/llpcElfWriter.cpp.o           
/ssd/opensource/amdvlk/vulkandriver/drivers/llpc/llpc/util/llpcElfWriter.cpp:1046:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  getSectionDataBySectionIndex(nonFragmentRelocSecIndex, &nonFragmentRelocSection);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ssd/opensource/amdvlk/vulkandriver/drivers/llpc/llpc/util/llpcElfWriter.cpp:1393:16: note: in instantiation of member function 'Llpc::ElfWriter<Vkgc::Elf64>::processRelocSection' requested 
here
template class ElfWriter<Elf64>;
               ^
/ssd/opensource/amdvlk/vulkandriver/drivers/llpc/llpc/util/llpcElfWriter.cpp:1069:5: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
    getSectionDataBySectionIndex(nonFragmentRelocSecIndex, &nonFragmentRelocSection);
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ssd/opensource/amdvlk/vulkandriver/drivers/llpc/llpc/util/llpcElfWriter.cpp:1216:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  getSectionDataBySectionIndex(nonFragmentSecIndex, &nonFragmentTextSection);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ssd/opensource/amdvlk/vulkandriver/drivers/llpc/llpc/util/llpcElfWriter.cpp:1393:16: note: in instantiation of member function 'Llpc::ElfWriter<Vkgc::Elf64>::mergeElfBinary' requested here
template class ElfWriter<Elf64>;
               ^
/ssd/opensource/amdvlk/vulkandriver/drivers/llpc/llpc/util/llpcElfWriter.cpp:1280:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  getSectionDataBySectionIndex(nonFragmentDisassemblySecIndex, &nonFragmentDisassemblySection); 
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ssd/opensource/amdvlk/vulkandriver/drivers/llpc/llpc/util/llpcElfWriter.cpp:1318:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  getSectionDataBySectionIndex(nonFragmentLlvmIrSecIndex, &nonFragmentLlvmIrSection);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/ssd/opensource/amdvlk/vulkandriver/drivers/llpc/llpc/util/llpcElfWriter.cpp:1148:3: warning: ignoring return value of function declared with 'warn_unused_result' attribute [-Wunused-result]
  getSectionDataBySectionIndex(textSectIdx, &textSection);
  ^~~~~~~~~~~~~~~~~~~~~~~~~~~~ ~~~~~~~~~~~~~~~~~~~~~~~~~
/ssd/opensource/amdvlk/vulkandriver/drivers/llpc/llpc/util/llpcElfWriter.cpp:1393:16: note: in instantiation of member function 'Llpc::ElfWriter<Vkgc::Elf64>::getSectionTextShader' requested
 here
template class ElfWriter<Elf64>;
```

`Result`s should either be checked or the return type should be changed.

**Comment by kuhar** [comment]
After https://github.com/GPUOpen-Drivers/llpc/pull/1483, there is one unhandled error that happens in `shaderdb/multiple_inputs/test_inputs/PipelineVsFs_ConstantData_Vs1Fs1.pipe`:

```
  auto fragmentLlvmIrSecIndex = reader.GetSectionIndex(llvmIrSectionName.c_str());
  auto nonFragmentLlvmIrSecIndex = GetSectionIndex(llvmIrSectionName.c_str());
  reader.getSectionDataBySectionIndex(fragmentLlvmIrSecIndex, &fragmentLlvmIrSection);
  result = getSectionDataBySectionIndex(nonFragmentLlvmIrSecIndex, &nonFragmentLlvmIrSection);
  (void)result;
```

**Comment by kuhar** [comment]
Another failure: `shaderdb/multiple_inputs/partial_pipeline_cache_test.multi-input`

```
void ElfWriter<Elf>::processRelocSection(const ElfReader<Elf> &reader, size_t nonFragmentPsIsaOffset,
                                         size_t fragmentPsIsaOffset) {
  SectionBuffer *fragmentRelocSection = nullptr;
  const SectionBuffer *nonFragmentRelocSection = nullptr;

  auto fragmentRelocSecIndex = reader.GetSectionIndex(RelocName);
  auto nonFragmentRelocSecIndex = GetSectionIndex(RelocName);

  reader.getSectionDataBySectionIndex(fragmentRelocSecIndex, &fragmentRelocSection);
  Result result = getSectionDataBySectionIndex(nonFragmentRelocSecIndex, &nonFragmentRelocSection);
```

cc: + @yukli-amd

**Comment by xuechen417** [comment]
> result = getSectionDataBySectionIndex(nonFragmentLlvmIrSecIndex, &nonFragmentLlvmIrSection);

**Comment by xuechen417** [comment]
> After #1483, there is one unhandled error that happens in `shaderdb/multiple_inputs/test_inputs/PipelineVsFs_ConstantData_Vs1Fs1.pipe`:
> 
> ```
>   auto fragmentLlvmIrSecIndex = reader.GetSectionIndex(llvmIrSectionName.c_str());
>   auto nonFragmentLlvmIrSecIndex = GetSectionIndex(llvmIrSectionName.c_str());
>   reader.getSectionDataBySectionIndex(fragmentLlvmIrSecIndex, &fragmentLlvmIrSection);
>   result = getSectionDataBySectionIndex(nonFragmentLlvmIrSecIndex, &nonFragmentLlvmIrSection);
assert(result == Result::Success || result == Result::ErrorInvalidValue); // Should it be handle like this?
>   (void)result;
> ```

**Comment by kuhar** [comment]
```
assert(result == Result::Success || result == Result::ErrorInvalidValue); // Should it be handle like this?
```

The purpose of check like this is to make sure that we don't miss edge cases or failures. I'm not very familiar with the code, so it's difficult for me to tell whether both statuses are fine and this is just an oversight, or whether this is a bug that needs to be fixed.

A proper fix should make explain why `ErrorInvalidValue` is fine there, handle this error value in some way, or abort if this is not a recoverable error.

**Comment by xuechen417** [comment]
> ```
> assert(result == Result::Success || result == Result::ErrorInvalidValue); // Should it be handle like this?
> ```
> 
> The purpose of check like this is to make sure that we don't miss edge cases or failures. I'm not very familiar with the code, so it's difficult for me to tell whether both statuses are fine and this is just an oversight, or whether this is a bug that needs to be fixed.
> 
> A proper fix should explain why `ErrorInvalidValue` is fine there, make sure we don't hit this error, handle this error value in some way, or abort if this is not a recoverable error.

I see. I think it's better for @yukli-amd to fix it.

**Comment by yukli-amd** [comment]
> > ```
> > assert(result == Result::Success || result == Result::ErrorInvalidValue); // Should it be handle like this?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The purpose of check like this is to make sure that we don't miss edge cases or failures. I'm not very familiar with the code, so it's difficult for me to tell whether both statuses are fine and this is just an oversight, or whether this is a bug that needs to be fixed.
> > A proper fix should explain why `ErrorInvalidValue` is fine there, make sure we don't hit this error, handle this error value in some way, or abort if this is not a recoverable error.
> 
> 
> I see. I think it's better for @yukli-amd to fix it.

@kuhar As I am not the original author of function `getSectionDataBySectionIndex`, but from my point of view, returning `ErrorInvalidValue` seems inappropriate here. 
Calling getSectionDataBySectionIndex is trying to find the section index, but the section may not exist, so I think it might make more sense to chang the function as follows.
```
template <class Elf>
Result ElfWriter<Elf>::getSectionDataBySectionIndex(unsigned secIdx, const SectionBuffer **ppSectionData) const {
  Result result = Result::NotFound;
  if (secIdx < m_sections.size()) {
    *ppSectionData = &m_sections[secIdx];
    result = Result::Success;
  }
  return result;
}
```
And then we can check the return value like,
`assert(result == Result::Success || result == Result::NotFound);`

Hope this can help.

**Comment by kuhar** [comment]
@yukli-amd Thanks for the explanation. Would it make more sense to make the function return `bool` instead then? Or `void` if nobody checks this result anyway?

**Comment by kuhar** [comment]
I updated the PR like you suggested: https://github.com/GPUOpen-Drivers/llpc/pull/1500. We can iterate on this and refine the API in a separate PR.

**Comment by yukli-amd** [comment]
> @yukli-amd Thanks for the explanation. Would it make more sense to make the function return `bool` instead then? Or `const SectionBuffer *` and do assignments outside this function?

From the usage, it may be better to return a `const SectionBuffer *`. I agree to refine this in a separate PR.

**Comment by kuhar** [comment]
I'm going to close this for now, as the API improvements seem orthogonal to the original issue.
